packages/plugin-cloudflare/src/tunnel.ts

```diff
@@ -0,0 +1,87 @@
+import {TUNNEL_PROVIDER} from './provider.js'
+import {startTunnel, TunnelError} from '@shopify/cli-kit/node/plugins/tunnel'
+import {err, ok, Result} from '@shopify/cli-kit/node/result'
+import {exec, captureOutput} from '@shopify/cli-kit/node/system'
+import {joinPath, dirname} from '@shopify/cli-kit/node/path'
+import {Writable} from 'stream'
+import {fileURLToPath} from 'url'
+
+export default startTunnel({provider: TUNNEL_PROVIDER, action: hookStart})
+
+export type ReturnType = Promise<Result<{url: string}, TunnelError>>
+
+export async function hookStart(port: number): ReturnType {
+  try {
+    const {url} = await tunnel({port})
+    return ok({url})
+    // eslint-disable-next-line no-catch-all/no-catch-all, @typescript-eslint/no-explicit-any
+  } catch (error: any) {
+    const tunnelError = new TunnelError('unknown', "Couldn't start tunnel")
+    return err(tunnelError)
+  }
+}
+
+async function tunnel(options: {port: number}): Promise<{url: string}> {
+  const args: string[] = ['tunnel', '--url', `http://localhost:${options.port}`, '--no-autoupdate']
+
+  let connected = false
+  let resolved = false
+  let url: string | undefined
+
+  return new Promise<{url: string}>((resolve, reject) => {
+    const customStdout = new Writable({
+      write(chunk, _, callback) {
+        if (resolved) return
+        if (!url) url = findUrl(chunk)
+        if (findConnection(chunk)) connected = true
+        if (url && connected) {
+          resolved = true
+          resolve({url})
+        }
+        if (findError(chunk)) reject(new Error("Couldn't start tunnel"))
+        callback()
+      },
+    })
+
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    exec(getBinPathTarget(), args, {
+      stdout: customStdout,
+      stderr: customStdout,
+    })
+  })
+}
+
+export async function cloudflareVersion(): Promise<string> {
+  return captureOutput(getBinPathTarget(), ['--version'])
+}
+
+function findUrl(data: Buffer): string | undefined {
+  const match = data.toString().match(/\|\s+(https?:\/\/[^\s]+)/) ?? undefined
```

Comment:
the magic of regex... it only matches urls that are inside a box 🙃 Just in case they change the format of the URL...
But if you prefer I can make it more explicit checking for trycloudflare.com

Reply:
Oh, now I see the `|` in the regex. If they change the output style it will break as well, so I think I prefer to check the current domain, at least the code would be clearer. But it's ok to leave it as it is.